synapse/rest/admin/room_hierarchy.py

```diff
@@ -0,0 +1,146 @@
+    """
+    Puppets a local user to remove them from all rooms in a space.
+    """
+
+    PATTERNS = admin_patterns(
+        "/rooms/(?P<space_id>[^/]+)/hierarchy/members/(?P<user_id>[^/]+)$"
+    )
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__(hs)
+        self._hs = hs
+        self._auth = hs.get_auth()
+        self._store = hs.get_datastore()
+        self._room_member_handler = hs.get_room_member_handler()
+        self._room_hierarchy_handler = hs.get_room_hierarchy_handler()
+
+    async def on_DELETE(
+        self, request: SynapseRequest, space_id: str, user_id: str
+    ) -> Tuple[int, JsonDict]:
+        """Forces a local user to leave all non-public rooms in a space.
+
+        The space itself is always left, regardless of whether it is public.
+
+        May succeed partially if the user fails to leave some rooms.
+
+        Returns:
+            A tuple containing the HTTP status code and a JSON dictionary containing:
+             * `left`: A list of rooms that the user has been made to leave.
+             * `failed`: A with entries for rooms that could not be fully processed.
+                The values of the dictionary are lists of failure reasons.
+                Rooms may appear here if:
+                 * The user failed to leave them for any reason.
+                 * The room is a space that the local homeserver is not in, and so its
+                   full list of child rooms could not be determined.
+                 * The room is inaccessible to the local homeserver, and it is not known
+                   whether the room is a subspace containing further rooms.
+                 * Some combination of the above.
+        """
+        requester = await self._auth.get_user_by_req(request)
+        await assert_user_is_admin(self._auth, requester.user)
+
+        space_id, _ = await self.resolve_room_id(space_id)
+
+        target_user = UserID.from_string(user_id)
+
+        if not self._hs.is_mine(target_user):
+            raise SynapseError(400, "This endpoint can only be used with local users")
+
+        # Fetch the list of rooms the target user is currently in
+        user_rooms = await self._store.get_rooms_for_local_user_where_membership_is(
+            user_id, [Membership.INVITE, Membership.JOIN, Membership.KNOCK]
+        )
+        user_room_ids = {room.room_id for room in user_rooms}
+
+        # Fetch the list of rooms in the space hierarchy
+        (
+            descendants,
+            inaccessible_room_ids,
+        ) = await self._room_hierarchy_handler.get_room_descendants(space_id)
+        space_room_ids = {space_id}
+        space_room_ids.update(room_id for room_id, _ in descendants)
+
+        # Determine which rooms to leave by checking join rules.
+        rooms_to_check = space_room_ids.intersection(user_room_ids)
+        rooms_to_leave = {space_id}  # Always leave the space, even if it is public
```

Comment:
Get rid of this maybe? But then it's a bit odd if an admin submits a request to remove a user from a space and Synapse does nothing.

Reply:
The code needs to check that `space_id` is in `user_room_ids`, otherwise it may try to leave a room that the user is not in.